ion-schema/src/result/invalid_schema_error.rs

```diff
@@ -0,0 +1,165 @@
+use crate::result::{HasIslSourceLocation, IonSchemaError, IonSchemaResult, IslSourceLocation};
+use std::collections::BTreeSet;
+use std::fmt::{Display, Formatter, Write};
+use thiserror::Error;
+
+/// Collects multiple errors to report them together.
+///
+/// ```ignore
+/// # use ion_schema::IonSchemaResult;
+/// #
+/// use ion_schema::result::ErrorCollector;
+///
+/// fn error_prone_function(i: i8) -> IonSchemaResult<()> {
+///     todo!()
+/// }
+/// fn some_fun() -> IonSchemaResult<()> {
+///     // Create the ErrorCollector
+///     let mut errors = ErrorCollector::default();
+///     for i in 1..10 {
+///         let Some(ok_value) = errors.ok_or_push_err(error_prone_function(i)) else {
+///             continue;
+///         }
+///         // Do something with `ok_value`
+///     }
+///     // Transform the error collector into a result.
+///     errors.into_result_with(())
+/// }
+/// ```
+#[derive(Debug, Default)]
+pub(crate) struct InvalidSchemaErrorCollector {
+    errors: BTreeSet<InvalidSchemaDetails>,
+}
+impl InvalidSchemaErrorCollector {
+    /// Consumes `self` and `ok_value`. If this `ErrorCollector` contains no errors, returns `Ok(ok_value)`.
+    /// Otherwise, returns `Err`.
+    ///
+    /// Use this method to turn the `ErrorCollector` into a usable `IonSchemaResult`.
+    pub(crate) fn into_result_with<T>(self, ok_value: T) -> Result<T, InvalidSchemaError> {
+        if self.errors.is_empty() {
+            Ok(ok_value)
+        } else {
+            Err(InvalidSchemaError(InvalidSchemaErrorQuantity::Many(
+                self.errors,
+            )))
+        }
+    }
+
+    /// Push an `IonSchemaError` to this `ErrorCollector`.
+    pub(crate) fn push_err(&mut self, e: impl Into<InvalidSchemaError>) {
+        let e = e.into();
+        match e.0 {
+            InvalidSchemaErrorQuantity::Many(errors) => self.errors.extend(errors),
+            InvalidSchemaErrorQuantity::One(error) => {
+                self.errors.insert(error);
+            }
+        }
+    }
+
+    /// Converts from [`IonSchemaResult<T>`] to [`Option<T>`].
+    ///
+    /// Converts `result` into an [`Option<T>`], consuming `result`, and pushing the error, if any,
+    /// to this `ErrorCollector`.
+    /// This is the multi-error equivalent to [`Result::ok`].
+    pub(crate) fn ok_or_push_err<T>(
+        &mut self,
+        result: impl Into<Result<T, InvalidSchemaError>>,
+    ) -> Option<T> {
+        match result.into() {
+            Ok(value) => Some(value),
+            Err(e) => {
+                self.push_err(e);
+                None
+            }
+        }
+    }
+}
+
+#[derive(Debug, Error, PartialEq)]
+#[error(transparent)]
+pub struct InvalidSchemaError(InvalidSchemaErrorQuantity);
```

Comment:
❗ An `InvalidSchemaError` is now composed of one or more `InvalidSchemaDetails`. `InvalidSchemaErrorQuantity` has a `One` and `Many` variant so that we can avoid allocating a `BTreeSet` unless we know we're collecting multiple errors. Heap allocation for a collection is _not_ required for the common case of invoking the `invalid_schema!` macro (although it may still be required for some of the `String`s in `InvalidSchemaDetails`).